### PR TITLE
Added support for external image proxy

### PR DESCRIPTION
**Pull request by albertlast**
By reading this: https://www.simplemachines.org/community/index.php?topic=558946.0
and see that different solution exists ( https://github.com/willnorris/imageproxy) which all need the data in same way:  proxy_url + image_url
And noticing that the amount of changes not high i decide to add this feature offical

What it do:
- add a second level of image proxy = 0 is disable 1 is proxy.php 2 is external proxy
- fix a docu issue $image_proxy_enabled is everywhere a int and not a bool
- add $image_proxy_url

external proxy works only when you switch the mode to 2 AND the $image_proxy_url is not empty otherwise the smf use proxy.php.

### CODE REVIEW

**Comment by Sesquipedalian** [comment]
I haven't tested, but this looks pretty good. Can I get you to adjust the admin interface in ManageServer.php's `ModifyGeneralSettings()` in order to support this?

**Comment by albertlast** [comment]
As expected the gui addition takes more time as the technical stuff,
reason why i hold this back until i got a go from your side.

File: Sources/Load.php [hunk 1]
```diff
@@ -1270,7 +1273,10 @@ function loadMemberData($users, $is_name = false, $set = 'normal')
 
 			// Take care of proxying avatar if required, do this here for maximum reach
 			if ($image_proxy_enabled && !empty($row['avatar']) && stripos($row['avatar'], 'http://') !== false && empty($user_info['possibly_robot']))
-				$row['avatar'] = $boardurl . '/proxy.php?request=' . urlencode($row['avatar']) . '&hash=' . md5($row['avatar'] . $image_proxy_secret);
+				if ($image_proxy_enabled === 2 && !empty($image_proxy_url))
+					$row['avatar'] = $image_proxy_url . $row['avatar'];
```
Comment (Sesquipedalian) [inline Sources/Load.php]:
```
$row['avatar'] = $image_proxy_url . urlencode($row['avatar']);
```

File: Themes/default/languages/Help.english.php [hunk 2]
```diff
@@ -618,9 +618,14 @@
...
-$helptxt['image_proxy_enabled'] = 'Required for embedding external images when in full SSL';
+$helptxt['image_proxy_enabled'] = 'Required for embedding external images when in full SSL: <br>
+		<ul class="normallist"> 
+			<li>Internal Image Proxy uses the internal proxy.php</li>
+			<li>External Image Proxy uses an external proxy service</li>
+		</ul>';
 $helptxt['image_proxy_secret'] = 'Keep this a secret, protects your forum from hotlinking images. Change it in order to render current hotlinked images useless';
 $helptxt['image_proxy_maxsize'] = 'Maximum image size that the SSL image proxy will cache: bigger images will be not be cached. Cached images are stored in your SMF cache folder, so make sure you have enough free space.';
+$helptxt['image_proxy_url'] = 'Needs to be maintained with the proxy url, when empty the internal proxy will be used';
```
Comment (Sesquipedalian) [inline Themes/default/languages/Help.english.php]: This needs some more descriptive help text. I suggest this:
```
Enter the URL for your external proxy <strong>exactly</strong> the way it should be accessed for the image file.<br><br>For example, if your proxy serves files via <code>http://www.example.com/yourproxy.php?file=[URL of proxied file]</code>, enter <code>http://www.example.com/yourproxy.php?file=</code> here.<br><br>If this field is left empty, SMF's internal proxy will be used.
```

File: Themes/default/languages/ManageSettings.english.php [hunk 3]
```diff
@@ -97,6 +97,10 @@
 $txt['force_ssl'] = 'Forum SSL mode';
 $txt['force_ssl_off'] = 'Disable SSL';
 $txt['force_ssl_complete'] = 'Force SSL throughout the forum';
+$txt['image_proxy_off'] = 'Disable';
+$txt['image_proxy_internal'] = 'Internal';
+$txt['image_proxy_external'] = 'External';
+$txt['image_proxy_url'] = 'URL of external Image Proxy';
```
Comment (Sesquipedalian) [inline Themes/default/languages/ManageSettings.english.php]: 'URL of external image proxy' ← lowercase

**Review by Sesquipedalian — CHANGES_REQUESTED** [review]
Two fixes and a suggestion for better help text.

File: Sources/Load.php [hunk 4]
```diff
@@ -446,7 +446,10 @@ function loadUserSettings()
 			$smcFunc['db_free_result']($request);
 
 			if (!empty($modSettings['force_ssl']) && $image_proxy_enabled && stripos($user_settings['avatar'], 'http://') !== false && empty($user_info['possibly_robot']))
-				$user_settings['avatar'] = strtr($boardurl, array('http://' => 'https://')) . '/proxy.php?request=' . urlencode($user_settings['avatar']) . '&hash=' . md5($user_settings['avatar'] . $image_proxy_secret);
+				if ($image_proxy_enabled === 2 && !empty($image_proxy_url))
+					$user_settings['avatar'] = $image_proxy_url . $user_settings['avatar'];
```
Comment (Sesquipedalian) [inline Sources/Load.php]:
```
$row['avatar'] = $image_proxy_url . urlencode($user_settings['avatar']);
```

**Review by Sesquipedalian — CHANGES_REQUESTED** [review]
One more requested change that I meant to include in the previous review.

**Comment by Sesquipedalian** [comment]
dc464c6 missed the other urlencode that we need. :)

**Comment by albertlast** [comment]
nope the others are already urlencode?

**Comment by Sesquipedalian** [comment]
[This one](https://github.com/SimpleMachines/SMF2.1/pull/4565/files#diff-e44bf29380cd171e044f5bf3ac967c8dR1277).

**Comment by albertlast** [comment]
okay i miss this one...
will i change later this day

**Comment by albertlast** [comment]
done

**Review by Sesquipedalian — APPROVED** [review]

**Comment by albertlast** [comment]
By thinking of that what we done,
was the urlencode a bad idea.
Why you wanted to urlencode @Sesquipedalian ?

**Pull request merged**